Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift

```diff
@@ -0,0 +1,284 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift OTel open source project
+//
+// Copyright (c) 2024 Moritz Lang and the Swift OTel project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import struct NIOConcurrencyHelpers.NIOLockedValueBox
+
+public final class OTelMetricRegistry: Sendable {
+    private let logger = Logger(label: "OTelMetricRegistry")
+
+    struct Storage {
+        var counters = [InstrumentIdentifier: [Set<Attribute>: Counter]]()
+        var gauges = [InstrumentIdentifier: [Set<Attribute>: Gauge]]()
+        var valueHistograms = [InstrumentIdentifier: [Set<Attribute>: ValueHistogram]]()
+        var durationHistograms = [InstrumentIdentifier: [Set<Attribute>: DurationHistogram]]()
+
+        var registrations = [String: Set<InstrumentIdentifier>]()
+        let duplicateRegistrationHandler: DuplicateRegistrationHandler?
+
+        mutating func register(_ identifier: InstrumentIdentifier, forName name: String) {
+            if var existingRegistrations = registrations[name] {
+                duplicateRegistrationHandler?.handle(newRegistration: identifier, existingRegistrations: existingRegistrations)
+                existingRegistrations.insert(identifier)
+                registrations[name] = existingRegistrations
+            } else {
+                registrations[name] = [identifier]
+            }
+        }
+
+        mutating func unregister(_ identifier: InstrumentIdentifier, forName name: String) {
+            guard var existingRegistrations = registrations[name] else { return }
+            existingRegistrations.remove(identifier)
+            if existingRegistrations.isEmpty {
+                registrations.removeValue(forKey: name)
+            }
+        }
+    }
+
+    let storage: NIOLockedValueBox<Storage>
+
+    public struct DuplicateRegistrationBehavior: Sendable {
+        enum Behavior: Sendable {
+            case warn, crash
+        }
+
+        var behavior: Behavior
+        public static let warn = Self(behavior: .warn)
+        public static let crash = Self(behavior: .crash)
+    }
+
+    internal init(duplicateRegistrationHandler: some DuplicateRegistrationHandler) {
+        self.storage = .init(Storage(duplicateRegistrationHandler: duplicateRegistrationHandler))
+    }
+
+    public convenience init(onDuplicateRegistration: DuplicateRegistrationBehavior = .warn) {
```

Comment:
Done. I didn't spell out that it defaults to `.warn` since that's the default parameter. Usually you only document the the default when the parameter is an optional that defaults to `nil` and you want to explain what happens when you use `nil`.

Reply:
Thanks, makes sense 👍